File: sample/src/main/java/me/xiaopan/assemblyadaptersample/fragment/GridRecyclerViewFragment.java
package me.xiaopan.assemblyadaptersample.fragment;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.format.Formatter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import me.xiaopan.assemblyadapter.AssemblyRecyclerAdapter;
import me.xiaopan.assemblyadaptersample.R;
import me.xiaopan.assemblyadaptersample.bean.AppInfo;
import me.xiaopan.assemblyadaptersample.itemfactory.AppItemFactory;
import me.xiaopan.assemblyadaptersample.itemfactory.AppListHeaderItemFactory;

public class GridRecyclerViewFragment extends Fragment {

    private AssemblyRecyclerAdapter adapter;
    private RecyclerView recyclerView;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_recycler_view, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        recyclerView = (RecyclerView) view.findViewById(R.id.list_recyclerViewFragment_content);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getActivity(), 4);
        gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                RecyclerView.Adapter adapter = recyclerView.getAdapter();
                return adapter != null && adapter instanceof AssemblyRecyclerAdapter
                        ? ((AssemblyRecyclerAdapter) adapter).getSpanSize(position) : 1;
            }
        });
        recyclerView.setLayoutManager(gridLayoutManager);

        if (adapter != null) {
            recyclerView.setAdapter(adapter);
        } else {
            loadAppList();
        }
    }

    private void loadAppList() {
        new AsyncTask<Integer, Integer, List<AppInfo>[]>() {
            private Context context = getActivity().getBaseContext();

            @Override
            protected void onPreExecute() {
                super.onPreExecute();
            }

            @Override
            protected List<AppInfo>[] doInBackground(Integer... params) {
                PackageManager packageManager = context.getPackageManager();
                List<PackageInfo> packageInfoList = packageManager.getInstalledPackages(0);
                List<AppInfo> systemAppList = new ArrayList<AppInfo>();
                List<AppInfo> userAppList = new ArrayList<AppInfo>();
                for (PackageInfo packageInfo : packageInfoList) {
                    AppInfo appInfo = new AppInfo(true);
                    appInfo.setName(String.valueOf(packageInfo.applicationInfo.loadLabel(packageManager)));
                    appInfo.setSortName(appInfo.getName());
                    appInfo.setId(packageInfo.packageName);
                    appInfo.setVersionName(packageInfo.versionName);
                    appInfo.setApkFilePath(packageInfo.applicationInfo.publicSourceDir);
                    appInfo.setAppSize(Formatter.formatFileSize(context, new File(appInfo.getApkFilePath()).length()));
                    appInfo.setVersionCode(packageInfo.versionCode);
                    if ((packageInfo.applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0) {
                        systemAppList.add(appInfo);
                    } else {
                        userAppList.add(appInfo);
                    }
                }

                Collections.sort(systemAppList, new Comparator<AppInfo>() {
                    @Override
                    public int compare(AppInfo lhs, AppInfo rhs) {
                        return lhs.getSortName().compareToIgnoreCase(rhs.getSortName());
                    }
                });

                return new List[]{systemAppList, userAppList};
            }

            @Override
            protected void onPostExecute(List<AppInfo>[] appInfoLists) {
                if (getActivity() == null) {
                    return;
                }

                List<AppInfo> systemAppList = appInfoLists[0];
                List<AppInfo> userAppList = appInfoLists[1];

                int systemAppListSize = systemAppList != null ? systemAppList.size() : 0;
                int userAppListSize = userAppList != null ? userAppList.size() : 0;

                int dataListSize = systemAppListSize > 0 ? systemAppListSize + 1 : 0;
                dataListSize += userAppListSize > 0 ? userAppListSize + 1 : 0;

                List<Object> dataList = new ArrayList<Object>(dataListSize);
                if (userAppListSize > 0) {
                    dataList.add(String.format("自安装应用%d个", userAppListSize));
                    dataList.addAll(userAppList);
                }
                if (systemAppListSize > 0) {
                    dataList.add(String.format("系统应用%d个", systemAppListSize));
                    dataList.addAll(systemAppList);
                }
                AssemblyRecyclerAdapter adapter = new AssemblyRecyclerAdapter(dataList);
                adapter.addItemFactory(new AppItemFactory());
                adapter.addItemFactory(new AppListHeaderItemFactory(recyclerView));
                recyclerView.setAdapter(adapter);
                recyclerView.scheduleLayoutAnimation();
                GridRecyclerViewFragment.this.adapter = adapter;
            }
        }.execute(0);
    }
}

File: sample/src/main/java/me/xiaopan/assemblyadaptersample/fragment/GridRecyclerViewFragment.kt
package me.xiaopan.assemblyadaptersample.fragment

import android.content.pm.ApplicationInfo
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.format.Formatter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import me.xiaopan.assemblyadapter.AssemblyRecyclerAdapter
import me.xiaopan.assemblyadaptersample.R
import me.xiaopan.assemblyadaptersample.bean.AppInfo
import me.xiaopan.assemblyadaptersample.itemfactory.AppItemFactory
import me.xiaopan.assemblyadaptersample.itemfactory.AppListHeaderItemFactory
import me.xiaopan.assemblyadaptersample.bindView
import java.io.File
import java.util.*

class GridRecyclerViewFragment : Fragment() {

    val recyclerView: RecyclerView by bindView(R.id.list_recyclerViewFragment_content)
    
    var adapter: AssemblyRecyclerAdapter? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_recycler_view, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val gridLayoutManager = GridLayoutManager(activity, 4)
        gridLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val adapter = recyclerView.adapter
                if (adapter == null || adapter !is AssemblyRecyclerAdapter) {
                    return 1
                }
                return adapter.getSpanSize(position)
            }
        }
        recyclerView.layoutManager = gridLayoutManager

        if (adapter != null) {
            recyclerView.adapter = adapter
        } else {
            loadAppList()
        }
    }

    private fun loadAppList() {
        object : AsyncTask<Int, Int, Array<List<AppInfo>>>() {
            private val context = activity.baseContext

            override fun onPreExecute() {
                super.onPreExecute()
            }

            override fun doInBackground(vararg params: Int?): Array<List<AppInfo>>? {
                val packageManager = context.packageManager
                val packageInfoList = packageManager.getInstalledPackages(0)
                val systemAppList = ArrayList<AppInfo>()
                val userAppList = ArrayList<AppInfo>()
                for (packageInfo in packageInfoList) {
                    val appInfo = AppInfo(true)
                    appInfo.name = packageInfo.applicationInfo.loadLabel(packageManager).toString()
                    appInfo.sortName = appInfo.name
                    appInfo.id = packageInfo.packageName
                    appInfo.versionName = packageInfo.versionName
                    appInfo.apkFilePath = packageInfo.applicationInfo.publicSourceDir
                    appInfo.appSize = Formatter.formatFileSize(context, File(appInfo.apkFilePath).length())
                    appInfo.versionCode = packageInfo.versionCode
                    if (packageInfo.applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM != 0) {
                        systemAppList.add(appInfo)
                    } else {
                        userAppList.add(appInfo)
                    }
                }

                Collections.sort(systemAppList) { lhs, rhs ->
                    (lhs.sortName ?: "").compareTo(rhs.sortName ?: "")
                }

                return arrayOf<List<AppInfo>>(systemAppList, userAppList)
            }

            override fun onPostExecute(appInfoLists: Array<List<AppInfo>>) {
                if (activity == null) {
                    return
                }

                val systemAppList = appInfoLists[0]
                val userAppList = appInfoLists[1]

                val systemAppListSize = systemAppList?.size ?: 0
                val userAppListSize = userAppList?.size ?: 0

                var dataListSize = if (systemAppListSize > 0) systemAppListSize + 1 else 0
                dataListSize += if (userAppListSize > 0) userAppListSize + 1 else 0

                val dataList = ArrayList<Any>(dataListSize)
                if (userAppListSize > 0) {
                    dataList.add(String.format("自安装应用%d个", userAppListSize))
                    dataList.addAll(userAppList)
                }
                if (systemAppListSize > 0) {
                    dataList.add(String.format("系统应用%d个", systemAppListSize))
                    dataList.addAll(systemAppList)
                }
                val adapter = AssemblyRecyclerAdapter(dataList)
                adapter.addItemFactory(AppItemFactory())
                adapter.addItemFactory(AppListHeaderItemFactory().fullSpan(recyclerView))
                recyclerView.adapter = adapter
                recyclerView.scheduleLayoutAnimation()
                this@GridRecyclerViewFragment.adapter = adapter
            }
        }.execute(0)
    }
}
